store/types/pruning.go

```diff
@@ -1,66 +1,74 @@
 package types
 
+import "fmt"
+
+// Pruning option string constants
+const (
+	PruningOptionDefault    = "default"
+	PruningOptionEverything = "everything"
+	PruningOptionNothing    = "nothing"
+	PruningOptionCustom     = "custom"
+)
+
 var (
-	// PruneEverything defines a pruning strategy where all committed states will
-	// be deleted, persisting only the current state.
-	PruneEverything = PruningOptions{
-		KeepEvery:     1,
-		SnapshotEvery: 0,
-	}
+	// PruneDefault defines a pruning strategy where the last 100 heights are kept
+	// in addition to every 500th and where to-be pruned heights are pruned at
+	// every 10th height.
+	PruneDefault = NewPruningOptions(100, 500, 10)
 
-	// PruneNothing defines a pruning strategy where all committed states will be
-	// kept on disk, i.e. no states will be pruned.
-	PruneNothing = PruningOptions{
-		KeepEvery:     1,
-		SnapshotEvery: 1,
-	}
+	// PruneEverything defines a pruning strategy where all committed heights are
+	// deleted, storing only the current height and where to-be pruned heights are
+	// pruned at every 10th height.
+	PruneEverything = NewPruningOptions(0, 0, 10)
 
-	// PruneSyncable defines a pruning strategy where only those states not needed
-	// for state syncing will be pruned. It flushes every 100th state to disk and
-	// keeps every 10000th.
-	PruneSyncable = PruningOptions{
-		KeepEvery:     100,
-		SnapshotEvery: 10000,
-	}
+	// PruneNothing defines a pruning strategy where all heights are kept on disk.
+	PruneNothing = NewPruningOptions(0, 1, 0)
 )
 
-// PruningOptions defines the specific pruning strategy every store in a multi-store
-// will use when committing state, where keepEvery determines which committed
-// heights are flushed to disk and snapshotEvery determines which of these heights
-// are kept after pruning.
+// PruningOptions defines the pruning strategy used when determining which
+// heights are removed from disk when committing state.
 type PruningOptions struct {
-	KeepEvery     int64
-	SnapshotEvery int64
+	// KeepRecent defines how many recent heights to keep on disk.
+	KeepRecent uint64
+
+	// KeepEvery defines how many offset heights are kept on disk past KeepRecent.
+	KeepEvery uint64
+
+	// Interval defines when the pruned heights are removed from disk.
+	Interval uint64
 }
 
-// IsValid verifies if the pruning options are valid. It returns false if invalid
-// and true otherwise. Pruning options are considered valid iff:
-//
-// - KeepEvery > 0
-// - SnapshotEvery >= 0
-// - SnapshotEvery % KeepEvery = 0
-func (po PruningOptions) IsValid() bool {
-	// must flush at positive block interval
-	if po.KeepEvery <= 0 {
-		return false
+func NewPruningOptions(keepRecent, keepEvery, interval uint64) PruningOptions {
+	return PruningOptions{
+		KeepRecent: keepRecent,
+		KeepEvery:  keepEvery,
+		Interval:   interval,
 	}
+}
 
-	// cannot snapshot negative intervals
-	if po.SnapshotEvery < 0 {
-		return false
+func (po PruningOptions) Validate() error {
```

Comment:
@colin-axner can you think of cases I've missed? I'd like to fix this on master too.

Reply:
Cases I can think of that are a bit funky because of the pruning must have an interval > 0:
- keep every != 0 (>1), interval = 0 (expects some pruning, no pruning happens)
- keep recent > 0, keep every = 0, interval = 0 (prune everything with some recents, but again nothing gets pruned)

Reply:
I think just dropping the `KeepRecent == 0 &&` (I don't see why KeepRecent would need to be checked) and changing the second if to use `>= 1` instead of `== 1`

Reply:
https://github.com/cosmos/cosmos-sdk/pull/6511